torch/fx/passes/graph_transform_observer.py

```diff
@@ -83,22 +84,41 @@ def _check_disable_pass(self):
         )
 
     def __enter__(self):
-        if self.log_url is None or self.gm is None:
-            return self
-
         self.erased_nodes = set()
         self.created_nodes = set()
+        self.replaced_nodes = {}
         self.gm._register_create_node_hook(self.on_node_creation)
         self.gm._register_erase_node_hook(self.on_node_erase)
+        self.gm._register_replace_node_hook(self.on_node_replace)
 
         return self
 
     def __exit__(self, type, value, tb):
-        if self.log_url is None or self.gm is None:
-            return
-
         self.gm._unregister_create_node_hook(self.on_node_creation)
         self.gm._unregister_erase_node_hook(self.on_node_erase)
+        self.gm._unregister_replace_node_hook(self.on_node_replace)
+
+        for node in self.gm.graph.nodes:
```

Comment:
Why do we need the following explicit propagation here? Does it mean "node.meta" is still not properly maintained at some places?

Reply:
> Why do we need the following explicit propagation here? Does it mean "node.meta" is still not properly maintained at some places?

In some sense, yes. The node meta is simply copied over from the old node at the time of node creation or node replacement. So here we update the node meta. 

An alternative is we change the create_node_hook and replace_node_hook so node meta is updated at the time of node creation/node replacement.

Reply:
I would prefer doing the propagation eagerly than lazily, in case anything in the middle of a pass needs the most up-to-date information.